### PR TITLE
Refactoring code

### PR DESCRIPTION
**Pull request by srenault**

### CODE REVIEW

File: src/Api.ts [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+  req?: any;
+  apiCache?: ApiCache;
+  apiDataTTL?: number;
+  proxyAgent?: any;
+}
+
+export default class Api {
+  url: string;
+  options: ApiOptions;
+  apiDataTTL: number;
```
Comment (arnaudlewis) [inline src/Api.ts]: deja present dans les options d'API

File: src/Api.ts [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  httpClient: HttpClient;
```
Comment (arnaudlewis) [inline src/Api.ts]: idem, on utilise les options d'api pour créer une instance et on utilise pas celle ci.

File: src/ApiData.ts [hunk 3]
```diff
@@ -0,0 +1,19 @@
+export interface QuickRoutes {
```
Comment (arnaudlewis) [inline src/ApiData.ts]: on va tout revoir les quickroutes, ce code ne sert pas on peut l'enlever.

File: src/ApiData.ts [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+  enabled: boolean;
+}
+
+export interface ApiData {
+  refs: Ref[];
+  bookmarks: { [key: string]: string };
+  types: { [key: string]: string };
+  tags: string[];
+  forms: { [key: string]: Form };
+  quickRoutes: QuickRoutes;
```
Comment (arnaudlewis) [inline src/ApiData.ts]: a enlever

File: src/ResolvedApi.ts [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+
+export interface Ref {
+  ref: string;
+  label: string;
+  isMasterRef: string;
+  scheduledAt: string;
+  id: string;
+}
+
+export interface ApiData {
```
Comment (arnaudlewis) [inline src/ResolvedApi.ts]: dommage de dupliquer ca. surtout que celui ci est clean alors que l'autre a les quickroutes. A unifier je pense.

File: src/ResolvedApi.ts [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+   * Returns the ref ID for a given ref's label.
+   * Do not use like this: searchForm.ref(api.ref("Future release label")).
+   * Instead, set your ref once in a variable, and call it when you need it; this will allow to change the ref you're viewing easily for your entire page.
+   */
+  ref(label: string): string | null {
+    const ref = this.data.refs.filter(ref => ref.label === label)[0];
+    return ref ? ref.ref : null;
+  }
+
+  currentExperiment(): Experiment | undefined {
```
Comment (arnaudlewis) [inline src/ResolvedApi.ts]: why undefined here instead of null like everywhere else?

File: src/Api.ts (hunk 2 above)
Comment (srenault) [reply src/Api.ts]: It's used on line 44 in `get` function

File: src/Api.ts (hunk 1 above)
Comment (srenault) [reply src/Api.ts]: this one is `this.options.apiDataTTL || 5;` (line 30)

File: src/ResolvedApi.ts (hunk 5 above)
Comment (srenault) [reply src/ResolvedApi.ts]: There is only `ApiData`. I don't see what you mean.

File: src/ResolvedApi.ts [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+  }
+
+  /**
+   * Retrieve the document returned by the given query
+   * @param {string|array|Predicate} the query
+   * @param {object} additional parameters. In NodeJS, pass the request as 'req'.
+   * @param {function} callback(err, doc)
+   */
+  queryFirst(q: string | string[], optionsOrCallback: QueryOptions | RequestCallback<Document>, cb?: RequestCallback<Document>): Promise<Document> {
+    const { options, callback } = typeof optionsOrCallback === 'function'
```
Comment (arnaudlewis) [inline src/ResolvedApi.ts]: Would be great to prevent doing this piece of code everywhere at some point.

File: src/ApiData.ts (hunk 4 above)
Comment (srenault) [reply src/ApiData.ts]: Okey

File: src/ApiData.ts (hunk 3 above)
Comment (srenault) [reply src/ApiData.ts]: Okey

File: src/ResolvedApi.ts (hunk 5 above)
Comment (srenault) [reply src/ResolvedApi.ts]: Are you comparing `ResolvedApi` and `ApiData` ?

File: src/ResolvedApi.ts (hunk 6 above)
Comment (srenault) [reply src/ResolvedApi.ts]: We could use `null` instead of `undefined`

File: src/ResolvedApi.ts (hunk 7 above)
Comment (srenault) [reply src/ResolvedApi.ts]: Does it means removing callback support?

File: src/Api.ts (hunk 1 above)
Comment (arnaudlewis) [reply src/Api.ts]: ah yes alright, I get it.

File: src/ResolvedApi.ts (hunk 7 above)
Comment (arnaudlewis) [reply src/ResolvedApi.ts]: maybe have a helper to do this or something less painful or yes maybe simply drop callback support I don't know yet what's the best

File: src/ResolvedApi.ts (hunk 5 above)
Comment (srenault) [reply src/ResolvedApi.ts]: okey

**Comment by srenault** [comment]
I think we shouldn't have the documentation files commited in the repository. It's a pain to regenerate documentation each times. We could publish the documentation from travis.

**Pull request merged**